src/objectExplorer/objectExplorerService.ts

```diff
@@ -275,17 +306,21 @@ export class ObjectExplorerService {
                     const sessionId = await this.createSession(promise, element.connectionCredentials);
                     if (sessionId) {
                         let node = await promise;
-                        // if password failed
+                        // if the server was found but connection failed
                         if (!node) {
-                            const connectNode = new ConnectTreeNode(element);
-                            this._treeNodeToChildrenMap.set(element, [connectNode]);
-                            return [connectNode];
+                            // remove password if it's saved in the credential store
+                            let profile = element.connectionCredentials as IConnectionProfile;
+                            let password = await this._connectionManager.connectionStore.lookupPassword(profile);
+                            if (password) {
+                                await this.removeProfilePassword(profile);
+                                return this.handleNodeCreationFailure(element);
+                            } else {
+                                return this.handleNodeConnectionFailure(element);
```

Comment:
I don't understand why you have these two node types based on whether a password was stored in the credential store. Isn't that something separate from whether a connection was successfully made? As mentioned in the other thread what about password-less connections such as integrated auth/AAD?

Reply:
Because we have 2 kinds of UX - one where only a password prompt is shown (was a login error) and one where the whole connection was unable to be made (server not found etc), in which case we go through the whole UX of recreating a profile with the same profile name.

Reply:
Ok - so it really isn't about the type of error that occurred is what I'm understanding. It's just that one flow is ONLY for re-entering a password to try again and the other is recreating the whole connection so is what you use for password-less connections.

The names here are really confusing and don't indicate that at all. "Node Creation Failure" and "Node Connection Failure" isn't really what's happening here at all - in both cases you failed to connect for some reason (which may or may not be related to a wrong password). I'd just call them "CreateSignInChildNode" and "CreateConnectChildNode"  or something to make it super clear what's happening and then explain in comments why you're doing the different ones here based on the password.

Reply:
Ok, so when you make a new Object Explorer session, you sent an OE request, which responds with a boolean on whether that create OE call was successful or not. That's the `NodeCreationFailure` part (if `createSession` has an `undefined` response). However, this is not all. We then receive another notification for the result node for the OE. This can fail too (wrong login, firewall etc). That's the `NodeConnectionFailure` case. In this case, `createSession` does have a response, but it fails in the notification handler. 

I have a comment here - https://github.com/microsoft/vscode-mssql/pull/1512/files#diff-9401975d07f4df26ce1df2d776b8c746R309

Reply:
How does this then get translated into the password existing in the store or not though? 

The call to createSession coming back with an undefined sessionId makes sense for what you're saying but I'm saying specifically here where you successfully made a session but the node itself wasn't created for some reason (as you said a connection error of some sort). 

Are you just reusing logic here or are these two situations (having a password stored vs not) actually somehow tied to whether a creation/connection failure occurred?

Reply:
The `handleNodeCreationFailure` case will give you a node under the connection, which when clicked will let you through the whole "Connection Retry" experience, which requires you to fill out the server, db etc all over again (since the connection was never made)

The `handleNodeConnectionFailure` case will give you a node under the connection, which when clicked will let you enter the password again (the only case I've noticed when we get an OE success response but OE notification failure). In all other cases, it failed in the initial create OE response itself. For the first create OE session call, it just looks at whether the connection details that are sent are null or not, and responds with a uri - https://github.com/microsoft/sqltoolsservice/blob/master/src/Microsoft.SqlTools.ServiceLayer/ObjectExplorer/ObjectExplorerService.cs#L182